keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,609 @@
+|----|-----|-----|-----|-----|-----|-----|-----|-----|
+| As | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 |
+| Aa | 1,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 |
+| Ab | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 |
+| Ba | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 |
+| Bb | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 |
+| Ca | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 |
+| Cb | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 | 0,0 |
+
+*Output suggestion*
+- expected (t/f)
+- result (T/F)
+- success / failed (1/0), pass = 1
+
+In the below matrix, both pods a and b could *successfully* connect to pod a (s), as expected (T), passing the test (1) 
+```      x/a
+x/a      sT1
+x/b      sT1 
+```
+  
+Most of the Matrices for this table will be permuting the first row and column, since the server pod currently always resides in the framework namespace.  However, tests might confirm two way connectivity
+and other types of connectivity in the future, and such an expansion would work very cleanly with a matrix.
+ 
+Part of this involves a pretty-print functionality for these tables, which can be output at the end of each Network policy test.  In failed
+test scenarios, these tables can be compared, and one may easily parse out a logical inference such as "Everything outside the framework
+namespace has connectivity, even when the truth table explicitly forbids it", which might, for example, point to a bug in a CNI provider
+related to flagrantly allowing internamespace traffic.  Since it is obvious how such a matrix might be defined in Go, we don't provide a
+code snippet or API example.
+ 
+3. (about 80% done modulo ginkgo implementation details) Rewrite each individual test, reviewing semantics, to be precisely worded (and possibly verbose), and to simply define a specific policy and
+set of 'whitelisted' communication associated with this policy. The whitelisting would be defined as a map of namespace->pods, since all other
+information in the truth table is false.
+ 
+Example:
+
+Initially, to confirm the logical capacity of the builder mechanism for replacing existing tests, a prototype of inplace replacements of NetworkPolicy definitions was done here (prototype) https://gist.github.com/6a62266e0eec2b15e5250bd65daa4faa.  
+Now, this underlying API has (essentially) been implemented in full, and the following repository https://github.com/vmware-tanzu/antrea/tree/master/hack/netpol, demonstrates a working implementation and port of the network policy tests (Most of the tests have been ported).  Each test follows a simple and easy to read pattern such as this:
+
+ ```
+	builder := &utils.NetworkPolicySpecBuilder{}
+	builder = builder.SetName("allow-x-via-pod-and-ns-selector").SetPodSelector(map[string]string{"pod": "a"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod":"b"}, map[string]string{"ns":"y"}, nil, nil)
+
+	reachability := utils.NewReachability(allPods, true)
```

Comment:
How about `reachability.New()` instead of an ambiguous `utils` package?

Reply:
+1, ill update